java/com/google/flatbuffers/FlexBuffersBuilder.java

```diff
@@ -0,0 +1,981 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.flatbuffers;
+
+import com.google.flatbuffers.FlexBuffers.BitWidth;
+
+import java.math.BigInteger;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+
+import static com.google.flatbuffers.FlexBuffers.*;
+import static com.google.flatbuffers.FlexBuffers.BitWidth.*;
+
+/**
+ * A class that generates FlexBuffers
+ * <p>
+ * This class presents all necessary APIs to create FlexBuffers. The {@link ByteBuffer } buffer used to store the
+ * data can be created internally, or passed down in the constructor.
+ * <p>
+ * Because it uses {@link ByteBuffer} internally, this impose some limitations in generating FlexBuffers. Mostly noted,
+ * the maximum size limitation on FlexBuffer message, which is {@link Integer#MAX_VALUE}.
+ *
+ * <p>There is also some differences from the original implementation in C++. It can changed in future updates.
+ * <ul>
+ *
+ *   <li><p>No support for mutations (might change in the future).</p></li>
+ *
+ *   <li><p>Size of message limited to {@link Integer#MAX_VALUE}</p></li>
+ *
+ *   <li><p>Since Java does not support unsigned type, all unsigned operations accepts a immediate higher representation
+ *   of similar type. Unsigned long is not supported</p></li>
+ * </ul>
+ * </p>
+ */
+public class FlexBuffersBuilder {
+    /**
+     * No keys or strings will be shared
+     */
+    public static final int BUILDER_FLAG_NONE = 0;
+    /**
+     * Keys will be shared between elements. Identical keys will only be serialized once, thus possibly saving space.
+     * But serialization performance might be slower and consumes more memory.
+     */
+    public static final int BUILDER_FLAG_SHARE_KEYS = 1;
+    /**
+     * Strings will be shared between elements. Identical strings will only be serialized once, thus possibly saving space.
+     * But serialization performance might be slower and consumes more memory. This is ideal if you expect many repeated
+     * strings on the message.
+     */
+    public static final int BUILDER_FLAG_SHARE_STRINGS = 1;
+    /**
+     * Strings and keys will be shared between elements.
+     */
+    public static final int BUILDER_FLAG_SHARE_KEYS_AND_STRINGS = 3;
+    /**
+     * Reserved for the future.
+     */
+    public static final int BUILDER_FLAG_SHARE_KEY_VECTORS = 4;
+    /**
+     * Reserved for the future.
+     */
+    public static final int BUILDER_FLAG_SHARE_ALL = 7;
+
+    private final ByteBuffer bb;
+    private final ArrayList<KeyValue> stack = new ArrayList<>();
```

Comment:
This is going to be an expensive way to track the stack, with each entry being 3 allocations. Maybe we can reduce it by either merging the key and the value objects (the key doesn't need type etc, since it is always the same), or by making the stack instead have 2 entries per stack element, where the key one can be `null` if its not a map.

Reply:
Turn the `Value key` into `int key`, so we should same some allocations

Reply:
Ok